desc/magnetic_fields.py

```diff
@@ -795,3 +831,464 @@ def odefun(rpz, s):
     r = x[:, :, 0].T.reshape((len(phis), *rshape))
     z = x[:, :, 2].T.reshape((len(phis), *rshape))
     return r, z
+
+
+class CurrentPotentialField(MagneticField, FourierRZToroidalSurface):
+    """Magnetic field due to a surface current potential on a toroidal surface.
+
+        surface current K is assumed given by
+         K = ∇ Φ x n
+        where n is the winding surface unit normal
+              Phi is the current potential function,
+              which is a function of theta and zeta
+        this function then uses biot-savart to find the
+        B field from this current density K on the surface
+
+        Note: potential function derivatives, if AD is used,
+        will not be vectorized over the params
+
+    Parameters
+    ----------
+    potential : callable
+        function to compute the current potential. Should have a signature of
+        the form potential(theta,zeta,**params) -> ndarray.
+        theta,zeta are poloidal and toroidal angles on the surface
+    surface_grid : Grid,
+        grid upon which to evaluate the surface current density K
+    params : dict, optional
+        default parameters to pass to potential function (and its derivatives)
+    potential_dtheta: callable
+        function to compute the theta derivative of the current potential
+        if None, will use AD to calculate
+    potential_dzeta: callable
+        function to compute the theta derivative of the current potential
+        if None, will use AD to calculate
+    R_lmn, Z_lmn : array-like, shape(k,)
+        Fourier coefficients for winding surface R and Z in cylindrical coordinates
+    modes_R : array-like, shape(k,2)
+        poloidal and toroidal mode numbers [m,n] for R_lmn.
+    modes_Z : array-like, shape(k,2)
+        mode numbers associated with Z_lmn, defaults to modes_R
+    NFP : int
+        number of field periods
+    sym : bool
+        whether to enforce stellarator symmetry. Default is "auto" which enforces if
+        modes are symmetric. If True, non-symmetric modes will be truncated.
+    rho : float [0,1]
+        flux surface label for the toroidal surface
+    name : str
+        name for this surface
+    check_orientation : bool
+        ensure that this surface has a right handed orientation. Do not set to False
+        unless you are sure the parameterization you have given is right handed
+        (ie, e_theta x e_zeta points outward from the surface).
+    surface: FourierRZToroidalSurface, optional, default None
```

Comment:
rather than passing this into init it might be better to make a `from_FourierRZSurface` method or `to_CurrentPotentialSurface` method in the surface class. Those methods would then only take the surface and potential functions, not all the rest of the stuff